nl_server/embeddings.py

```diff
@@ -80,52 +131,196 @@ def get_embedding_at_index(self, index: int) -> List[float]:
+
+    query2result: Dict[str, VarCandidates] = {}
+
+    for q, sv2score in query2sv2score.items():
+      sv2score_sorted = [(k, v) for (
+          k,
+          v) in sorted(sv2score.items(), key=lambda item: item[1], reverse=True)
+                        ]
+      svs = [k for (k, _) in sv2score_sorted]
+      scores = [v for (_, v) in sv2score_sorted]
+      query2result[q] = VarCandidates(svs=svs, scores=scores, sv2sentences={})
+
+    for q, sv2sentence2score in query2sv2sentence2score.items():
+      query2result[q].sv2sentences = {}
+      for sv, sentence2score in sv2sentence2score.items():
+        query2result[q].sv2sentences[sv] = []
+        for sentence, score in sorted(sentence2score.items(),
+                                      key=lambda item: item[1],
+                                      reverse=True):
+          score = round(score, 4)
+          query2result[q].sv2sentences[sv].append(sentence + f' ({score})')
+
+    return query2result
+
+  #
+  # The main entry point to detect SVs.
+  #
+  def detect_svs(self, orig_query: str) -> Dict[str, Union[Dict, List]]:
+    # Remove all stop-words.
+    query_monovar = utils.remove_stop_words(orig_query,
+                                            query_util.ALL_STOP_WORDS)
+
+    # Search embeddings for a single SV.
+    result_monovar = self._search_embeddings([query_monovar])[query_monovar]
+
+    # Try to detect multiple SVs.  Use the original query so that
+    # the logic can rely on stop-words like `vs`, `and`, etc as hints
+    # for SV delimiters.
+    result_multivar = self._detect_multiple_svs(orig_query)
 
     return {
-        'SV': svs_sorted,
-        'CosineScore': scores_sorted,
-        'EmbeddingIndex': sv_index_sorted,
-        'SV_to_Sentences': all_svs_sentences,
+        'SV': result_monovar.svs,
+        'CosineScore': result_monovar.scores,
+        'SV_to_Sentences': result_monovar.sv2sentences,
+        'MultiSV': _multivar_candidates_to_dict(result_multivar)
+    }
+
+  #
+  # Detects one or more SVs from the query.
+  # TODO: Fix the query upstream to ensure the punctuations aren't stripped.
+  #
+  def _detect_multiple_svs(self, query: str) -> MultiVarCandidates:
+    #
+    # Prepare a combination of query-sets.
+    #
+    querysets = query_util.prepare_multivar_querysets(query)
+
+    result = MultiVarCandidates(candidates=[], sv2sentences={})
+
+    # Make a unique list of query strings
+    all_queries = set()
+    for qs in querysets:
+      for c in qs.combinations:
+        for p in c.parts:
+          all_queries.add(p)
+    if not all_queries:
+      return result
+
+    query2result = self._search_embeddings(list(all_queries))
+
+    #
+    # We take the average score from the top SV from the query-parts in a
+    # queryset (ignoring any queryset with a score below threshold). Then
+    # sort all candidates by that score.
+    #
+    # TODO: Come up with a better ranking function.
+    #
+    for qs in querysets:
```

Comment:
Some examples of "querysets" is helpful, since its fields are very deep

Reply:
Done, PTAL.